snowpackjs/snowpack
# Fix undefined accessor errors using @snowpack/plugin-babel with snowpack 3.1.x

## Changes

Previously, #649 added a hardcoded transform to `@snowpack/plugin-babel` that replaced all code instances of `process.env` with `import.meta.env`. After upgrading to snowpack 3.1.x, this change results in runtime errors when using `@snowpack/plugin-babel` (see #2979 and #2978).

The error seems to stem from a combination of package compilation changes in snowpack 3.1 and the way that this manual transform modifies package code. The end result is that `import.meta.env.SOME_PROP` is injected into package code (e.g. React initializer) and results in a runtime error since `import.meta.env` is `undefined` for package files.

This change adds a shim to the top of package files that are transformed by `@snowpack/plugin-babel`, as suggested in #2979:

```javascript
import.meta.env = import.meta.env || process.env || {};
```

There is no change for source files as snowpack already handles import.meta.env setup for them. I believe this is the optimal fix since the issue only arises when using `@snowpack/plugin-babel`.

## Testing

I added two unit tests for `@snowpack/plugin-babel` to:

1. Check that `process.env` is transformed to `import.meta.env` for source files, and
2. Check that the `import.meta.env` shim is added to package files

Additionally, I tested this locally to ensure it fixes the undefined access exception. I also tested with a `@babel/preset-env` config to make sure it's still getting loaded properly, and it is.

## Docs

No change to user-facing docs as this is a bug fix.
